File: snake_maze.js
// snake_maze.js — Змейка в лабиринте на JavaScript (Node.js)

const readline = require('readline');
const { EventEmitter } = require('events');

const WIDTH = 20, HEIGHT = 15;
const WALL = '#', EMPTY = ' ', HEAD = '@', BODY = 'o', FOOD = '$', BONUS = '*';

let maze = [];
let snake = [];
let food = {x:0,y:0}, bonus = null;
let bonusTimer = 0;
let dx = 1, dy = 0;
let score = 0, highScore = 0, level = 1;
let speed = 0.15;
let gameOver = false, paused = false;
let lastMove = Date.now();
let inputQueue = [];

// Чтение ввода без ожидания Enter (используем keypress)
readline.emitKeypressEvents(process.stdin);
process.stdin.setRawMode(true);
process.stdin.on('keypress', (str, key) => {
    if (key) {
        if (key.name === 'up') inputQueue.push('up');
        else if (key.name === 'down') inputQueue.push('down');
        else if (key.name === 'left') inputQueue.push('left');
        else if (key.name === 'right') inputQueue.push('right');
        else if (key.name === 'p') inputQueue.push('pause');
        else if (key.name === 'q') inputQueue.push('quit');
    }
});

function clearScreen() { console.clear(); }

function loadHighScore() { highScore = 0; }
function saveHighScore() {}

function generateMaze() {
    maze = Array.from({length: HEIGHT}, () => Array(WIDTH).fill(EMPTY));
    for (let x=0; x<WIDTH; x++) { maze[0][x] = WALL; maze[HEIGHT-1][x] = WALL; }
    for (let y=0; y<HEIGHT; y++) { maze[y][0] = WALL; maze[y][WIDTH-1] = WALL; }
    for (let y=2; y<HEIGHT-2; y++) {
        for (let x=2; x<WIDTH-2; x++) {
            if (Math.random() < 0.15) maze[y][x] = WALL;
        }
    }
}

function spawnSnake() {
    const cx = Math.floor(WIDTH/2), cy = Math.floor(HEIGHT/2);
    snake = [];
    for (let i=0; i<3; i++) snake.push({x: cx-i, y: cy});
    dx = 1; dy = 0;
}

function placeFood() {
    while (true) {
        const x = Math.floor(Math.random() * (WIDTH-2)) + 1;
        const y = Math.floor(Math.random() * (HEIGHT-2)) + 1;
        const p = {x,y};
        const inSnake = snake.some(s => s.x===x && s.y===y);
        if (!inSnake && maze[y][x] !== WALL && (!bonus || bonus.x!==x || bonus.y!==y)) {
            food = p;
            break;
        }
    }
}

function placeBonus() {
    if (Math.random() < 0.2) {
        while (true) {
            const x = Math.floor(Math.random() * (WIDTH-2)) + 1;
            const y = Math.floor(Math.random() * (HEIGHT-2)) + 1;
            const p = {x,y};
            const inSnake = snake.some(s => s.x===x && s.y===y);
            if (!inSnake && maze[y][x] !== WALL && (food.x!==x || food.y!==y)) {
                bonus = p;
                bonusTimer = 10;
                break;
            }
        }
    } else {
        bonus = null;
        bonusTimer = 0;
    }
}

function move() {
    if (gameOver || paused) return;
    const head = snake[0];
    const nx = head.x + dx;
    const ny = head.y + dy;
    if (nx<0 || nx>=WIDTH || ny<0 || ny>=HEIGHT || maze[ny][nx] === WALL) {
        gameOver = true;
        if (score > highScore) highScore = score;
        return;
    }
    const newHead = {x:nx, y:ny};
    if (snake.some(s => s.x===nx && s.y===ny)) {
        gameOver = true;
        return;
    }
    snake.unshift(newHead);
    let ate = false;
    if (nx===food.x && ny===food.y) {
        score++; ate = true;
        placeFood(); placeBonus();
        if (score%5===0) { level++; speed = Math.max(0.05, speed*0.9); }
    } else if (bonus && nx===bonus.x && ny===bonus.y) {
        score += 5; ate = true;
        bonus = null; bonusTimer = 0;
        placeFood(); placeBonus();
        if (score%5===0) { level++; speed = Math.max(0.05, speed*0.9); }
    }
    if (!ate) snake.pop();
    if (bonus) {
        bonusTimer--;
        if (bonusTimer<=0) bonus = null;
    }
}

function draw() {
    clearScreen();
    console.log(`🐍 SnakeMaze  |  Счёт: ${score}  |  Уровень: ${level}  |  Рекорд: ${highScore}`);
    if (paused) console.log("⏸ ПАУЗА");
    const top = '+' + '-'.repeat(WIDTH) + '+';
    console.log(top);
    for (let y=0; y<HEIGHT; y++) {
        let line = '|';
        for (let x=0; x<WIDTH; x++) {
            let c = maze[y][x];
            if (x===snake[0].x && y===snake[0].y) c = HEAD;
            else if (snake.some(s => s.x===x && s.y===y)) c = BODY;
            else if (food.x===x && food.y===y) c = FOOD;
            else if (bonus && bonus.x===x && bonus.y===y) c = BONUS;
            line += c;
        }
        line += '|';
        console.log(line);
    }
    console.log(top);
    console.log("Управление: стрелки, P - пауза, Q - выход");
}

function processInput() {
    while (inputQueue.length > 0) {
        const cmd = inputQueue.shift();
        if (cmd === 'quit') process.exit(0);
        if (cmd === 'pause') { paused = !paused; continue; }
        if (cmd === 'up' && dy !== 1) { dx=0; dy=-1; }
        else if (cmd === 'down' && dy !== -1) { dx=0; dy=1; }
        else if (cmd === 'left' && dx !== 1) { dx=-1; dy=0; }
        else if (cmd === 'right' && dx !== -1) { dx=1; dy=0; }
    }
}

function gameLoop() {
    if (!gameOver) {
        processInput();
        const now = Date.now();
        if ((now - lastMove) / 1000 > speed) {
            move();
            lastMove = now;
            if (gameOver) {
                draw();
                console.log(`ИГРА ОКОНЧЕНА! Счёт: ${score}`);
                saveHighScore();
                process.stdin.setRawMode(false);
                process.exit(0);
            }
        }
        draw();
        setTimeout(gameLoop, 20);
    }
}

// Инициализация
loadHighScore();
generateMaze();
spawnSnake();
placeFood();
placeBonus();
lastMove = Date.now();
gameLoop();

File: snake_maze.rs
// snake_maze.rs — Змейка в лабиринте на Rust

use std::io::{self, Write, stdout};
use std::thread;
use std::time::{Duration, Instant};
use rand::Rng;
use termion::{clear, cursor, color, style};
use termion::input::TermRead;

const WIDTH: usize = 20;
const HEIGHT: usize = 15;
const WALL: char = '#';
const EMPTY: char = ' ';
const HEAD: char = '@';
const BODY: char = 'o';
const FOOD: char = '$';
const BONUS: char = '*';

type Point = (usize, usize);

struct SnakeMaze {
    maze: Vec<Vec<char>>,
    snake: Vec<Point>,
    food: Point,
    bonus: Option<Point>,
    bonus_timer: i32,
    dir: (i32, i32),
    next_dir: (i32, i32),
    score: i32,
    high_score: i32,
    level: i32,
    speed: f64,
    game_over: bool,
    paused: bool,
}

impl SnakeMaze {
    fn new() -> Self {
        let mut game = SnakeMaze {
            maze: vec![vec![EMPTY; WIDTH]; HEIGHT],
            snake: Vec::new(),
            food: (0,0),
            bonus: None,
            bonus_timer: 0,
            dir: (1,0),
            next_dir: (1,0),
            score: 0,
            high_score: 0,
            level: 1,
            speed: 0.15,
            game_over: false,
            paused: false,
        };
        game.load_high_score();
        game.generate_maze();
        game.spawn_snake();
        game.place_food();
        game.place_bonus();
        game
    }

    fn load_high_score(&mut self) {
        self.high_score = 0;
    }

    fn generate_maze(&mut self) {
        for y in 0..HEIGHT {
            for x in 0..WIDTH {
                self.maze[y][x] = EMPTY;
            }
        }
        for x in 0..WIDTH {
            self.maze[0][x] = WALL;
            self.maze[HEIGHT-1][x] = WALL;
        }
        for y in 0..HEIGHT {
            self.maze[y][0] = WALL;
            self.maze[y][WIDTH-1] = WALL;
        }
        let mut rng = rand::thread_rng();
        for y in 2..HEIGHT-2 {
            for x in 2..WIDTH-2 {
                if rng.gen_range(0..100) < 15 {
                    self.maze[y][x] = WALL;
                }
            }
        }
    }

    fn spawn_snake(&mut self) {
        let cx = WIDTH/2;
        let cy = HEIGHT/2;
        self.snake.clear();
        for i in 0..3 {
            self.snake.push((cx - i, cy));
        }
        self.dir = (1,0);
        self.next_dir = (1,0);
    }

    fn place_food(&mut self) {
        let mut rng = rand::thread_rng();
        loop {
            let x = rng.gen_range(1..WIDTH-1);
            let y = rng.gen_range(1..HEIGHT-1);
            let p = (x,y);
            if !self.snake.contains(&p) && self.maze[y][x] != WALL {
                if let Some(b) = self.bonus {
                    if b == p { continue; }
                }
                self.food = p;
                break;
            }
        }
    }

    fn place_bonus(&mut self) {
        let mut rng = rand::thread_rng();
        if rng.gen_range(0..100) < 20 {
            loop {
                let x = rng.gen_range(1..WIDTH-1);
                let y = rng.gen_range(1..HEIGHT-1);
                let p = (x,y);
                if !self.snake.contains(&p) && self.maze[y][x] != WALL && self.food != p {
                    self.bonus = Some(p);
                    self.bonus_timer = 10;
                    break;
                }
            }
        } else {
            self.bonus = None;
            self.bonus_timer = 0;
        }
    }

    fn move_snake(&mut self) {
        if self.game_over || self.paused { return; }
        let head = self.snake[0];
        let nx = head.0 as i32 + self.dir.0;
        let ny = head.1 as i32 + self.dir.1;
        if nx < 0 || nx >= WIDTH as i32 || ny < 0 || ny >= HEIGHT as i32 {
            self.game_over = true;
            if self.score > self.high_score { self.high_score = self.score; }
            return;
        }
        let new_head = (nx as usize, ny as usize);
        if self.maze[new_head.1][new_head.0] == WALL {
            self.game_over = true;
            return;
        }
        if self.snake.contains(&new_head) {
            self.game_over = true;
            return;
        }
        self.snake.insert(0, new_head);
        let mut ate = false;
        if new_head == self.food {
            self.score += 1;
            ate = true;
            self.place_food();
            self.place_bonus();
            if self.score % 5 == 0 {
                self.level += 1;
                self.speed = (self.speed * 0.9).max(0.05);
            }
        } else if let Some(b) = self.bonus {
            if new_head == b {
                self.score += 5;
                ate = true;
                self.bonus = None;
                self.bonus_timer = 0;
                self.place_food();
                self.place_bonus();
                if self.score % 5 == 0 {
                    self.level += 1;
                    self.speed = (self.speed * 0.9).max(0.05);
                }
            }
        }
        if !ate {
            self.snake.pop();
        }
        if let Some(_) = self.bonus {
            self.bonus_timer -= 1;
            if self.bonus_timer <= 0 {
                self.bonus = None;
            }
        }
    }

    fn draw(&self) {
        print!("{}{}", clear::All, cursor::Goto(1,1));
        println!("🐍 SnakeMaze  |  Счёт: {}  |  Уровень: {}  |  Рекорд: {}", self.score, self.level, self.high_score);
        if self.paused {
            println!("⏸ ПАУЗА");
        }
        let top = format!("+{}+", "-".repeat(WIDTH));
        println!("{}", top);
        for y in 0..HEIGHT {
            print!("|");
            for x in 0..WIDTH {
                let mut c = self.maze[y][x];
                if x == self.snake[0].0 && y == self.snake[0].1 {
                    c = HEAD;
                } else if self.snake.iter().any(|p| p.0==x && p.1==y) {
                    c = BODY;
                } else if self.food.0==x && self.food.1==y {
                    c = FOOD;
                } else if let Some(b) = self.bonus {
                    if b.0==x && b.1==y { c = BONUS; }
                }
                print!("{}", c);
            }
            println!("|");
        }
        println!("{}", top);
        println!("Управление: стрелки, P - пауза, Q - выход");
        stdout().flush().unwrap();
    }

    fn run(&mut self) {
        let stdin = io::stdin();
        let mut keys = stdin.keys();
        let mut last_time = Instant::now();
        while !self.game_over {
            self.draw();
            // обработка ввода
            if let Some(Ok(key)) = keys.next() {
                match key {
                    termion::event::Key::Up => if self.dir.1 != 1 { self.next_dir = (0,-1); },
                    termion::event::Key::Down => if self.dir.1 != -1 { self.next_dir = (0,1); },
                    termion::event::Key::Left => if self.dir.0 != 1 { self.next_dir = (-1,0); },
                    termion::event::Key::Right => if self.dir.0 != -1 { self.next_dir = (1,0); },
                    termion::event::Key::Char('p') => self.paused = !self.paused,
                    termion::event::Key::Char('q') => return,
                    _ => {}
                }
            }
            self.dir = self.next_dir;
            if last_time.elapsed().as_secs_f64() > self.speed {
                self.move_snake();
                last_time = Instant::now();
                if self.game_over {
                    self.draw();
                    println!("ИГРА ОКОНЧЕНА! Счёт: {}", self.score);
                    break;
                }
            }
            thread::sleep(Duration::from_millis(20));
        }
    }
}

fn main() {
    let mut game = SnakeMaze::new();
    game.run();
}
